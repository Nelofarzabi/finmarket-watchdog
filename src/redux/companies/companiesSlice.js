import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const initialState = {
  companies: [],
  filteredCompanies: [],
  companyDetail: {},
  statement: [],
  loading: false,
  error: '',
  typing: false,
};
// const API_KEY = 'c0405893fd96337e82d58e3ee28f1c63';
const API_KEY = 'd0627ca47e8ec703f960f6fa495a3c45';

export const getActiveCompanies = createAsyncThunk('companies/getCompanies', async () => {
  try {
    const response = await fetch(`https://financialmodelingprep.com/api/v3/stock_market/actives?apikey=${API_KEY}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
});
export const getCompanyDetail = createAsyncThunk('companies/getCompanyDetail', async (symbol) => {
  try {
    const response = await fetch(`https://financialmodelingprep.com/api/v3/profile/${symbol}?apikey=${API_KEY}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
});
export const getCompanyStatement = createAsyncThunk('companies/getCompanyStatement', async (symbol) => {
  try {
    const response = await fetch(`https://financialmodelingprep.com/api/v3/income-statement/${symbol}?limit=120&apikey=${API_KEY}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return error;
  }
});

const companiesSlice = createSlice({
  name: 'companies',
  initialState,
  reducers: {
    filterCompanies: (state, action) => {
      state.filteredCompanies = state.companies
        .filter((company) => company.name.toLowerCase()
          .startsWith(action.payload.toLowerCase()));
    },
    updateTyping: (state, action) => {
      state.typing = !!action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getActiveCompanies.pending, (state) => {
        state.loading = true;
      })
      .addCase(getActiveCompanies.fulfilled, (state, action) => {
        state.loading = false;
        if (action.payload['Error Message'] || action.payload.message) {
          state.error = action.payload.message;
          state.companies = [];
        }
        state.companies = action.payload;
      })
      .addCase(getActiveCompanies.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
      .addCase(getCompanyDetail.pending, (state) => {
        state.loading = true;
      })
      .addCase(getCompanyDetail.fulfilled, (state, action) => {
        state.loading = false;
        [state.companyDetail] = action.payload;
      })
      .addCase(getCompanyDetail.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
      .addCase(getCompanyStatement.pending, (state) => {
        state.loading = false;
      })
      .addCase(getCompanyStatement.fulfilled, (state, action) => {
        state.loading = false;
        state.statement = action.payload;
      })
      .addCase(getCompanyStatement.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error;
      });
  },
});

export const { filterCompanies, updateTyping } = companiesSlice.actions;
export default companiesSlice.reducer;
