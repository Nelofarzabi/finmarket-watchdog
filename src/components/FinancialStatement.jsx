import { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getCompanyStatement } from '../redux/companies/companiesSlice';

const FinancialStatement = () => {
  const { symbol } = useParams();
  const dispatch = useDispatch();
  const { statement } = useSelector((state) => state.companies);
  useEffect(() => {
    dispatch(getCompanyStatement(symbol));
  }, [dispatch, symbol]);

  return (
    <div>
      <table className="w-full">
        <tbody>
          <tr>
            <th className="w-[10%]">Year</th>
            <th>Income(net)</th>
            <th>Expenses(operating)</th>
            <th>Gross Profit</th>
          </tr>
          {
            statement.map((state) => {
              const {
                calendarYear: year,
                grossProfit: profit,
                netIncome: income,
                operatingExpenses: expenses,
              } = state;
              return (
                <tr key={symbol}>
                  <td className="w-[10%]">{year}</td>
                  <td className="w-[10%]">
                    $
                    {income.toLocaleString()}
                  </td>
                  <td className="whitespace-nowrap w-[10%]">
                    $
                    {expenses.toLocaleString()}
                  </td>
                  <td className="w-[10%]">
                    $
                    {profit.toLocaleString()}
                  </td>
                </tr>
              );
            })
          }
        </tbody>
      </table>
    </div>
  );
};

export default FinancialStatement;
