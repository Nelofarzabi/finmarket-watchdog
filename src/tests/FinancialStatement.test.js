import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getCompanyStatement } from '../redux/companies/companiesSlice';
import FinancialStatement from '../components/FinancialStatement';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../redux/companies/companiesSlice', () => ({
  getCompanyStatement: jest.fn(),
}));

describe('FinancialStatement', () => {
  const mockDispatch = jest.fn();
  const mockStatement = [
    {
      calendarYear: 2021,
      grossProfit: 1000000,
      netIncome: 800000,
      operatingExpenses: 200000,
    },
    {
      calendarYear: 2020,
      grossProfit: 950000,
      netIncome: 750000,
      operatingExpenses: 200000,
    },
  ];

  beforeEach(() => {
    useDispatch.mockReturnValue(mockDispatch);
    useSelector.mockImplementation((selector) => selector({
      companies: {
        statement: mockStatement,
      },
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('dispatches getCompanyStatement action with the symbol parameter', async () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={['/statement/XYZ']}>
        <FinancialStatement />
      </MemoryRouter>,
    );

    // Ensure the component dispatches the action with the symbol parameter
    await waitFor(() => expect(mockDispatch).toHaveBeenCalledWith(getCompanyStatement('XYZ')));

    // Check if the component renders the statement data
    expect(getByText('2021')).toBeInTheDocument();
    expect(getByText('$1,000,000')).toBeInTheDocument();
    expect(getByText('$800,000')).toBeInTheDocument();
    expect(getByText('2020')).toBeInTheDocument();
    expect(getByText('$950,000')).toBeInTheDocument();
    expect(getByText('$750,000')).toBeInTheDocument();
  });

  test('matches snapshot', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/statement/XYZ']}>
        <FinancialStatement />
      </MemoryRouter>,
    );

    await waitFor(() => expect(mockDispatch).toHaveBeenCalledWith(getCompanyStatement('XYZ')));

    // Check if the rendered component matches the snapshot
    expect(container).toMatchSnapshot();
  });
});
