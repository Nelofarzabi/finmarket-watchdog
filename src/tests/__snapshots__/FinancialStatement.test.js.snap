// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinancialStatement matches snapshot 1`] = `
<div>
  <div>
    <table
      class="w-full md:w-[60%] mx-auto"
    >
      <tbody>
        <tr>
          <th
            class="text-center"
          >
            Year
          </th>
          <th
            class="text-center"
          >
            Income(net)
          </th>
          <th
            class="text-center"
          >
            Expenses(operating)
          </th>
          <th
            class="text-center"
          >
            Gross Profit
          </th>
        </tr>
        <tr>
          <td
            class="w-[10%]"
          >
            2021
          </td>
          <td
            class="w-[10%]"
          >
            $
            800,000
          </td>
          <td
            class="whitespace-nowrap w-[10%]"
          >
            $
            200,000
          </td>
          <td
            class="w-[10%]"
          >
            $
            1,000,000
          </td>
        </tr>
        <tr>
          <td
            class="w-[10%]"
          >
            2020
          </td>
          <td
            class="w-[10%]"
          >
            $
            750,000
          </td>
          <td
            class="whitespace-nowrap w-[10%]"
          >
            $
            200,000
          </td>
          <td
            class="w-[10%]"
          >
            $
            950,000
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
